presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/ColumnMapping.java

```diff
@@ -14,22 +14,53 @@
 package io.prestosql.plugin.jdbc;
 
 import io.prestosql.plugin.jdbc.PredicatePushdownController.DomainPushdownResult;
+import io.prestosql.spi.predicate.DiscreteValues;
 import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.Ranges;
 import io.prestosql.spi.type.Type;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.plugin.jdbc.JdbcMetadataSessionProperties.getDomainSizeThreshold;
 import static java.util.Objects.requireNonNull;
 
 public final class ColumnMapping
 {
-    public static final PredicatePushdownController FULL_PUSHDOWN = domain -> new DomainPushdownResult(domain, Domain.all(domain.getType()));
-    public static final PredicatePushdownController PUSHDOWN_AND_KEEP = domain -> new DomainPushdownResult(domain, domain);
-    public static final PredicatePushdownController DISABLE_PUSHDOWN = domain -> new DomainPushdownResult(Domain.all(domain.getType()), domain);
+    public static PredicatePushdownController fullPushDown()
+    {
+        return (session, domain) -> {
+            if (getDomainSize(domain) > getDomainSizeThreshold(session)) {
+                // pushdown simplified domain
+                return new DomainPushdownResult(domain.simplify(), domain);
+            }
+            else {
+                // full pushdown
+                return new DomainPushdownResult(domain, Domain.all(domain.getType()));
+            }
+        };
+    }
+
+    public static PredicatePushdownController pushDownAndKeep()
+    {
+        return (session, domain) -> new DomainPushdownResult(domain.simplify(getDomainSizeThreshold(session)), domain);
+    }
+
+    public static PredicatePushdownController disablePushdown()
+    {
+        return (session, domain) -> new DomainPushdownResult(Domain.all(domain.getType()), domain);
+    }
+
+    private static int getDomainSize(Domain domain)
+    {
+        return domain.getValues().getValuesProcessor().transform(
+                Ranges::getRangeCount,
```

Comment:
perhaps count ranges twice, when both bounds finite.. except that it won't play reasonably with simplify()